Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -0,0 +1,498 @@
+                        let (promise, providerMustClose) = self.stateLock.withLock { self.state.popConnectionPromiseToFail() }
+                        promise?.fail(error)
+                        if providerMustClose {
+                            self.parentPool[self.key] = nil
+                        }
+                    }
+                }
+
+            case .removeProvider:
+                // FIXME: Need to close the pool
+                break
+
+            case .none:
+                break
+            }
+        }
+
+        private func makeConnection(on eventLoop: EventLoop) -> EventLoopFuture<Connection> {
+            let bootstrap = ClientBootstrap.makeHTTPClientBootstrapBase(group: eventLoop, host: self.key.host, port: self.key.port, configuration: self.configuration) { channel in
+                channel.pipeline.addSSLHandlerIfNeeded(for: self.key, tlsConfiguration: self.configuration.tlsConfiguration).flatMap {
+                    channel.pipeline.addHTTPClientHandlers(leftOverBytesStrategy: .forwardBytes)
+                }
+            }
+            let address = HTTPClient.resolveAddress(host: self.key.host, port: self.key.port, proxy: self.configuration.proxy)
+            return bootstrap.connect(host: address.host, port: address.port).map { channel in
+                let connection = Connection(key: self.key, channel: channel, parentPool: self.parentPool)
+                self.configureCloseCallback(of: connection)
+                connection.isLeased.store(true)
+                return connection
+            }
+        }
+
+        /// Adds a callback on connection close that asks the `state` what to do about this
+        ///
+        /// The callback informs the state about the event, and the state returns a
+        /// `ClosedConnectionRemoveAction` which instructs it about what it should do.
+        private func configureCloseCallback(of connection: Connection) {
+            connection.channel.closeFuture.whenComplete { result in
+                switch result {
+                case .success:
+                    let action = self.stateLock.withLock { self.state.removeClosedConnection(connection) }
+                    switch action {
+                    case .none:
+                        break
+                    case .removeProvider:
+                        self.parentPool[self.key] = nil
+                    case .makeConnectionAndComplete(let el, let promise):
+                        self.makeConnection(on: el).cascade(to: promise)
+                    }
+
+                case .failure(let error):
+                    preconditionFailure("Connection close future failed with error: \(error)")
+                }
+            }
+        }
+
+        func syncClose(requiresCleanClose: Bool) throws {
+            let availableConnections = try self.stateLock.withLock { () -> CircularBuffer<ConnectionPool.Connection> in
+                if requiresCleanClose {
```

Comment:
I don't think this is quite right. I'm pretty sure we need to maintain some state here that we're closing. Because otherwise we're handing back the _currently_ `availableConnections` but that may change any time, right? For example if anybody were to request another connection whilst we're closing, that'd be a problem, no?

Reply:
@weissi I don't know if that's a correct way to reason about it, but when I look at the `Task.succeed` method, I see that I only succeed the task result future after the connection is released (ie: I release the connection inside of one of the `flatMap`s callbacks that lead to the result future being succeeded), so AFAICT when the user has all tasks finished, we can be sure that no more task will be released, and so the only reason why a new task would be released would be that the user made an new request after calling `syncClose` which would be their fault and will be catched by new checks I must add. What do you think?

Reply:
@adtrevor 

> and so the only reason why a new task would be released would be that the user made an new request after calling syncClose which would be their fault and will be catched by new checks I must add. What do you think?

That’s exactly the scenario I mean. We need to detect this situation and that’s why I think we need to store state here. I think we should just fail all requests attempted after syncClose. We cannot just have it do random behaviour as that’ll get us very tricky to repro bug reports where we have no idea what’s going on.

Reply:
I’d advise the pool to have one

```
enum State {
    case upAndRunning
    case shuttingDown
    case shutDown
}
```

And in `deinit` you should `assert(self.state == .shutDown)`

Once syncClose is called we should check that state is .upAndRunning and transition the state to .shuttingDown.

If a user makes a request in .shuttingDown or .shutDown we should fail it

Reply:
@weissi What would be the use of `shuttingDown`/`shutDown` vs the `self.isShutdown` boolean that already exists?

Reply:
@adtrevor well, there are 3 states:

1. up and running
2. we started the shutdown
3. we completed the shutdown

All three states are important:

- methods that do stuff with the pool, make sure that we're in (1)
- methods that initiate shutdown transition from (1) to (2)
- in deinit we make sure we're in (3)

the transition from (1) to (2) is important because otherwise if the user initiates shutdown twice in a row, you might do conflicting actions because you haven't remembered that you already running through the shutdown sequence.

Your existing boolean can only express either
- (1) (`false`) and (2) (`true`) which would be an issue in `deinit` because you can't know if you actually finished a shutdown
- (1) (`false`) and (3) (`true`) which would be an init in the shutdown methods because you can't know if you already started a shutdown

Reply:
@weissi  Thanks, the latest push should fix this, let me know if something can be improved